### ISSUE TITLE
Using Naive Quantizer for Speed_up.

### ISSUE THREAD

**Issue by YangNuoCheng** [microsoft/nni#4404]
## NNI 学生项目问题概述 / General Question of Student Program
Using Naive Quantizer for Speed_up.
**请简要概述您的问题 / 观点 :**
**Short summary about the question / idea :**
我该怎么使用Naive Quantizer做加速？已知在class NaiveQuantizer(Quantizer)并没有继承export_model等功能，我无法直接将它和Speedup模块相连，我无法得到[nni文档](https://nni.readthedocs.io/zh/latest/Compression/QuantizationSpeedup.html)中的加速效果，是否有其他的案例可以使用？
我使用QAT和tensorrt加速时，"all in 8 bit"和mixed precision(average bit 20.4)速度相差不大，这是什么原因呢？我的GPU架构不支持加速功能吗？
| quantization strategy      | Latency | accuracy     |
| :----:       |    :----:   |         :----:  |
| all in 32bit      | 0.001199961       | 96%   |
| mixed precision(average bit 20.4)   | 0.000753688        | 96%     |
|all in 8bit | 0.000753688        | 93.7%     |
**请提供 NNI 环境信息 :**
**nni Environment :**
- NNI version: 2.5
- Training service (local|remote|pai|aml|etc): remote
- Client OS: Ubuntu 18.04
- Server OS (for remote mode only): Ubuntu18
- Python version: 3.9
- PyTorch/TensorFlow version: 1.6
- Is conda/virtualenv/venv used?: Yes
- Is running in Docker?: No
- GPU TITAN RTX
- CUDA 11.5
- Tensorrt 8.2.1.8

## 其他建议 / Other Advice

**是否需要更新文档（是 / 否）:**
**Need to update document ( yes / no ) :**
yes
**其他分享内容 :**
**Anything else we need to know :**

**Log message / 日志信息 :**
 - [nnimanager.log and dispatcher.log](https://github.com/microsoft/nni/blob/master/docs/en_US/Tutorial/HowToDebug.md#experiment-root-directory) : 

 - [nnictl stdout and stderr](https://github.com/microsoft/nni/blob/master/docs/en_US/Tutorial/Nnictl.md#nnictl%20log%20stdout) :

**Comment by J-shang** [comment]
`NaiveQuantizer` doesn't support speed up, it just simulated to see the effect of if we directly quantize the weight.

FYI, I think your GPU support int8, https://docs.nvidia.com/deeplearning/tensorrt/support-matrix/index.html#hardware-precision-matrix

@linbinskn, please help to explain the `mixed precision`.

**Comment by YangNuoCheng** [comment]
> `NaiveQuantizer` doesn't support speed up, it just simulated to see the effect of if we directly quantize the weight.
> 
> FYI, I think your GPU support int8, https://docs.nvidia.com/deeplearning/tensorrt/support-matrix/index.html#hardware-precision-matrix
> 
> @linbinskn, please help to explain the `mixed precision`.

Thank you for your reply!
I am concerned about whether I use different `config_list`  really quantify the model to different type. 
Because I observed that their accuracy, loss and latency variations were very similar.
Do I misused the config_list of QAT?

**Comment by linbinskn** [comment]
Your gpu card supports low precision very well since Titan RTX is boosted by turing architecture which leverages tensorcore low precision speedup.
I think config_list is right in terms of format and it may really work. The result in document is model accuracy after post-training quantization whose accuracy should be lower than quantize aware training. Actually, mnist is a very simple task so that qat should have similar accuracy to pretrained-model. 
There are several reasons for why they have same speedup result. The most possible one influential may be only some parts of network are quantized while others are keeping original precision. Maybe you can try the following config to see the further result.
`configure_list = [{
        'quant_types': ['weight', 'input'],
        'quant_bits': {'weight': 8, 'input': 8},
        'op_names': ['conv1', 'conv2']
    }, {
        'quant_types': ['output'],
        'quant_bits': {'output': 8, },
        'op_names': ['relu1', 'relu2']
    }, {
        'quant_types': ['output', 'weight', 'input'],
        'quant_bits': {'output': 8, 'weight': 8, 'input': 8},
        'op_names': ['fc1', 'fc2'],
    }]`

**Comment by YangNuoCheng** [comment]
## Thanks for your reply! ##

You solved a lot of my problems！
But I still got the following results:
|`all in 8bit`|`all in 16bit`|
| :----:       |    :----:   |
| 0.034036874771118164s|0.03283333778381348s|
| 0.03186678886413574s|0.032286882400512695s|
|0.03266406059265137s|0.03229546546936035s|
|0.032309532165527344s|0.03174257278442383s|
|0.032884836196899414s|0.03217005729675293s|

which  `all in 8bit` still similar to `all in 16bit`, are there other reasons led to speedup result?
<details>
  <summary>here are some detials `output`</summary>
  <p> 
NaiveModel(
  (conv1): Conv2d(1, 20, kernel_size=(5, 5), stride=(1, 1))
  (conv2): Conv2d(20, 50, kernel_size=(5, 5), stride=(1, 1))
  (fc1): Linear(in_features=800, out_features=500, bias=True)
  (fc2): Linear(in_features=500, out_features=10, bias=True)
  (relu1): ReLU6()
  (relu2): ReLU6()
  (relu3): ReLU6()
  (max_pool1): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
  (max_pool2): MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
)
# Epoch 0 #
 0%  Loss 2.320648670196533
11%  Loss 1.5353659391403198
21%  Loss 0.5524840950965881
32%  Loss 0.2670537531375885
43%  Loss 0.15946835279464722
53%  Loss 0.23263391852378845
64%  Loss 0.30145013332366943
75%  Loss 0.06855656951665878
85%  Loss 0.04781820625066757
96%  Loss 0.21670007705688477
Loss: 0.09943255615234375  Accuracy: 96.9%)

[2021-12-22 21:55:41] INFO (nni.compression.pytorch.compressor/MainThread) Model state_dict saved to mnist_model.pth
[2021-12-22 21:55:41] INFO (nni.compression.pytorch.compressor/MainThread) Mask dict saved to mnist_calibration.pth
calibration_config 
{'conv1': {'weight_bits': 16, 'weight_scale': tensor([9.5225e-06], device='cuda:0'), 'weight_zero_point': tensor([26380.], device='cuda:0'), 'input_bits': 16, 'tracked_min_input': -0.4242129623889923, 'tracked_max_input': 2.821486711502075}, 'conv2': {'weight_bits': 16, 'weight_scale': tensor([3.4201e-06], device='cuda:0'), 'weight_zero_point': tensor([30375.], device='cuda:0'), 'input_bits': 16, 'tracked_min_input': 0.0, 'tracked_max_input': 5.998416423797607}, 'fc1': {'weight_bits': 16, 'weight_scale': tensor([1.6138e-06], device='cuda:0'), 'weight_zero_point': tensor([32241.], device='cuda:0'), 'input_bits': 16, 'tracked_min_input': 0.0, 'tracked_max_input': 5.987966537475586, 'output_bits': 16, 'tracked_min_output': -5.204525947570801, 'tracked_max_output': 7.360141277313232}, 'fc2': {'weight_bits': 16, 'weight_scale': tensor([4.0696e-06], device='cuda:0'), 'weight_zero_point': tensor([32722.], device='cuda:0'), 'input_bits': 16, 'tracked_min_input': 0.0, 'tracked_max_input': 5.899948596954346, 'output_bits': 16, 'tracked_min_output': -18.205446243286133, 'tracked_max_output': 8.411518096923828}, 'relu1': {'output_bits': 16, 'tracked_min_output': 0.0, 'tracked_max_output': 5.9998369216918945}, 'relu2': {'output_bits': 16, 'tracked_min_output': 0.0, 'tracked_max_output': 5.998589992523193}, 'relu3': {'output_bits': 16, 'tracked_min_output': 0.0, 'tracked_max_output': 5.983097553253174}}
Loss: 0.09942604522705079  Accuracy: 96.9%)

[12/22/2021-21:55:43] [TRT] [W] onnx2trt_utils.cpp:366: Your ONNX model has been generated with INT64 weights, while TensorRT does not natively support INT64. Attempting to cast down to INT32.
[12/22/2021-21:55:43] [TRT] [W] DynamicRange(min: -0.424213, max: 2.82149). Dynamic range should be symmetric for better accuracy.
[12/22/2021-21:55:43] [TRT] [W] DynamicRange(min: -5.20453, max: 7.36014). Dynamic range should be symmetric for better accuracy.
[12/22/2021-21:55:43] [TRT] [W] DynamicRange(min: -18.2054, max: 8.41152). Dynamic range should be symmetric for better accuracy.
[12/22/2021-21:55:43] [TRT] [W] Calibrator is not being used. Users must provide dynamic range for all tensors that are not Int32 or Bool.
[12/22/2021-21:55:43] [TRT] [W] Missing scale and zero-point for tensor 12, expect fall back to non-int8 implementation for any layer consuming or producing given tensor
[12/22/2021-21:55:43] [TRT] [W] Missing scale and zero-point for tensor 21, expect fall back to non-int8 implementation for any layer consuming or producing given tensor
[12/22/2021-21:55:43] [TRT] [W] Missing scale and zero-point for tensor 27, expect fall back to non-int8 implementation for any layer consuming or producing given tensor
[12/22/2021-21:55:43] [TRT] [W] Missing scale and zero-point for tensor (Unnamed Layer* 14) [Shuffle]_output, expect fall back to non-int8 implementation for any layer consuming or producing given tensor
[12/22/2021-21:55:43] [TRT] [W] Missing scale and zero-point for tensor (Unnamed Layer* 15) [Softmax]_output, expect fall back to non-int8 implementation for any layer consuming or producing given tensor
[12/22/2021-21:55:43] [TRT] [W] Missing scale and zero-point for tensor (Unnamed Layer* 16) [Unary]_output, expect fall back to non-int8 implementation for any layer consuming or producing given tensor
[12/22/2021-21:55:43] [TRT] [W] Missing scale and zero-point for tensor output1, expect fall back to non-int8 implementation for any layer consuming or producing given tensor
<class 'tensorrt.tensorrt.ICudaEngine'>
Loss: 0.09943007965087891  Accuracy: 96.91%
Inference elapsed_time (whole dataset): 0.03217005729675293s
</p>
  <pre><code>  title，value，callBack可以缺省  </code>  </pre>
</details>

<details>
  <summary>here are some detials of `quantization_aware_training_example`</summary>
  <p> 
def quantization_aware_training_example(train_loader, test_loader, device):
    model = NaiveModel()
    bits = 16
    configure_list = [
        { 'quant_types': ['weight', 'input'], 
         'quant_bits': {'weight': bits, 'input': bits}, 
         'op_names': ['conv1', 'conv2'] 
        }, 
        { 'quant_types': ['output'], 
         'quant_bits': {'output': bits }, 
         'op_names': ['relu1', 'relu2','relu3'] 
        }, 
        { 'quant_types': ['output', 'weight', 'input'], 
         'quant_bits': {'output': bits, 'weight': bits, 'input': bits},
         'op_names': ['fc1', 'fc2'], }]
    print(model)
    # set_quant_scheme_dtype('weight', 'per_tensor_symmetric', 'int')
    # set_quant_scheme_dtype('input', 'per_tensor_symmetric', 'int')
    # set_quant_scheme_dtype('output', 'per_tensor_symmetric', 'int')
    # finetune the model by using QAT
    # enable batchnorm folding mode
    dummy_input = torch.randn(1, 1, 28, 28)
    optimizer = torch.optim.SGD(model.parameters(), lr=0.01, momentum=0.5)
    quantizer = QAT_Quantizer(model, configure_list, optimizer, dummy_input=dummy_input)
    # quantizer = QAT_Quantizer(model, configure_list, optimizer)
    quantizer.compress()

    model.to(device)
    for epoch in range(1):
        print('# Epoch {} #'.format(epoch))
        train(model, device, train_loader, optimizer)
        test(model, device, test_loader)

    model_path = "mnist_model.pth"
    calibration_path = "mnist_calibration.pth"
    calibration_config = quantizer.export_model(model_path, calibration_path)

    test(model, device, test_loader)

    # print("calibration_config: ", calibration_config)

    batch_size = 32
    input_shape = (batch_size, 1, 28, 28)

    engine = ModelSpeedupTensorRT(model, input_shape, config=calibration_config, batchsize=batch_size)
    engine.compress()

    test_trt(engine, test_loader)
</p>
  <pre><code>  title，value，callBack可以缺省  </code>  </pre>
</details>
